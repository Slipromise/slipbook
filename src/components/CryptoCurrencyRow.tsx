"use client";

import Image from "react-bootstrap/esm/Image";
import styles from "@/styles/components/CryptoCurrencyRow.module.scss";
import { useSpring, animated } from "@react-spring/web";
import { useMemo } from "react";

type Props = {
  name: string;
  rank: number;
  symbol: string;
  symbolUri: string;
  currency: string;
  price: number;
  changeIn1h: number;
  changeIn24h: number;
  changeIn1w: number;
  volume24h: number;
  marketCap: number;
  onSubscribe: () => void;
  onClick: () => void;
  locale?: string;
};

function CryptoCurrencyRow({
  name,
  rank,
  symbol,
  symbolUri,
  currency,
  price,
  changeIn1h,
  changeIn24h,
  changeIn1w,
  volume24h,
  marketCap,
  // onSubscribe,
  // onClick,
  locale = "en-US",
}: Props) {
  const animateValues = useSpring({
    price,
    changeIn1h,
    changeIn24h,
    changeIn1w,
    marketCap,
    volume24h,
    config: {
      duration: 300,
    },
  });

  const numberFormat = useMemo(
    () =>
      new Intl.NumberFormat(locale, {
        style: "currency",
        currency,
        notation: "compact",
        compactDisplay: "short",
        maximumSignificantDigits: 2,
      }),
    [currency, locale]
  );

  const volumeFormat = useMemo(
    () =>
      new Intl.NumberFormat(locale, {
        style: "percent",
        signDisplay: "exceptZero",
      }),
    [locale]
  );

  return (
    <tr className={styles.container}>
      <td data-rank>
        <span>{rank}</span>
      </td>
      <td data-name>
        <Image src={symbolUri} alt="Crypto Currency" />
        <span>{name}</span>
        <span>{symbol}</span>
      </td>
      <td data-price>
        <animated.span>
          {animateValues.price.to((n) => numberFormat.format(n))}
        </animated.span>
      </td>
      <td data-changeIn1h>
        <animated.span
          data-positive={changeIn1h > 0}
          data-negative={changeIn1h < 0}
        >
          {animateValues.changeIn1h.to((n) => volumeFormat.format(n / 100))}
        </animated.span>
      </td>
      <td data-changeIn24h>
        <animated.span
          data-positive={changeIn24h > 0}
          data-negative={changeIn24h < 0}
        >
          {animateValues.changeIn24h.to((n) => volumeFormat.format(n / 100))}
        </animated.span>
      </td>
      <td data-changeIn1w>
        <animated.span
          data-positive={changeIn1w > 0}
          data-negative={changeIn1w < 0}
        >
          {animateValues.changeIn1w.to((n) => volumeFormat.format(n / 100))}
        </animated.span>
      </td>
      <td data-marketCap>
        <animated.span>
          {animateValues.marketCap.to((n) => numberFormat.format(n))}
        </animated.span>
      </td>
      <td data-volume24h>
        <animated.span>
          {animateValues.volume24h.to((n) => numberFormat.format(n))}
        </animated.span>
      </td>
    </tr>
  );
}

export default CryptoCurrencyRow;

export const CryptoCurrencyHeaderRow = () => {
  return (
    <tr>
      <th data-rank>#</th>
      <th data-name>Name</th>
      <th data-price>Price</th>
      <th data-changeIn1h>1h%</th>
      <th data-changeIn24h>24h%</th>
      <th data-changeIn1w>7d%</th>
      <th data-marketCap>Market Cap</th>
      <th data-volume24h>Volume(24h)</th>
    </tr>
  );
};
