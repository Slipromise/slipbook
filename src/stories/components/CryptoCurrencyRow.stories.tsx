import type { Meta, StoryObj } from "@storybook/react";
import React, { useEffect } from "react";
import CryptoCurrencyRow, {
  CryptoCurrencyHeaderRow,
} from "../../components/CryptoCurrencyRow";
import Table from "react-bootstrap/Table";
import { run as runHolder } from "holderjs";

// TODO: 組合式Story

const meta: Meta<typeof CryptoCurrencyRow> = {
  component: CryptoCurrencyRow,
  tags: ["autodocs"],
  decorators: [
    (story) => {
      const StoryComponent = story;

      // TODO: 無法使用withHolder

      useEffect(() => {
        runHolder();
      }, []);
      return (
        <Table striped bordered hover>
          <thead>
            <CryptoCurrencyHeaderRow />
          </thead>
          <tbody>
            <StoryComponent />
            <StoryComponent />
          </tbody>
        </Table>
      );
    },
  ],
};

export default meta;

type Stroy = StoryObj<typeof CryptoCurrencyRow>;

export const Normal: Stroy = {
  args: {
    name: "bitcoin",
    rank: 1,
    symbol: "BTC",
    symbolUri: "holder.js/32x32?auto=yes&random=yes",
    price: 51730.160788641166,
    pricingUnit: "USD",
    changeIn1h: -0.18975577,
    changeIn24h: -0.13245235,
    changeIn1w: 9.5045419,
    marketCap: 1015473040202.0583,
    volume24h: 23436245752.813572,
  },
};
