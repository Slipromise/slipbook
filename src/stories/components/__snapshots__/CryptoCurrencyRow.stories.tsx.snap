// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/CryptoCurrencyRow BNB smoke-test 1`] = `
<tr class="_container_y2li0_1">
  <td data-rank="true">
    <span>
      4
    </span>
  </td>
  <td data-name="true">
    <img src="data:image/svg+xml;charset=UTF-8,%3Csvg%20width%3D%2232%22%20height%3D%2232%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%2032%2032%22%20preserveAspectRatio%3D%22none%22%3E%3Cdefs%3E%3Cstyle%20type%3D%22text%2Fcss%22%3E%23holder_18f57f0a593%20text%20%7B%20fill%3A%23FFFFFF%3Bfont-weight%3Abold%3Bfont-family%3AArial%2C%20Helvetica%2C%20Open%20Sans%2C%20sans-serif%2C%20monospace%3Bfont-size%3A10pt%20%7D%20%3C%2Fstyle%3E%3C%2Fdefs%3E%3Cg%20id%3D%22holder_18f57f0a593%22%3E%3Crect%20width%3D%2232%22%20height%3D%2232%22%20fill%3D%22%233a5a97%22%3E%3C%2Frect%3E%3Cg%3E%3Ctext%20x%3D%22-2.5390625%22%20y%3D%2228%22%3E32x32%3C%2Ftext%3E%3C%2Fg%3E%3C%2Fg%3E%3C%2Fsvg%3E"
         alt="Crypto Currency"
         class
    >
    <span>
      BNB
    </span>
    <span>
      BNB
    </span>
  </td>
  <td data-price="true">
    <span>
      $421.07
    </span>
  </td>
  <td data-changein1h="true">
    <span data-positive="false"
          data-negative="true"
    >
      -0.18%
    </span>
  </td>
  <td data-changein24h="true">
    <span data-positive="false"
          data-negative="true"
    >
      -0.13%
    </span>
  </td>
  <td data-changein1w="true">
    <span data-positive="true"
          data-negative="false"
    >
      9.5%
    </span>
  </td>
  <td data-marketcap="true">
    <span>
      $3.04M
    </span>
  </td>
  <td data-volume24h="true">
    <span>
      $45.75K
    </span>
  </td>
</tr>
`;

exports[`stories/components/CryptoCurrencyRow Bitcoin smoke-test 1`] = `
<tr class="_container_y2li0_1">
  <td data-rank="true">
    <span>
      1
    </span>
  </td>
  <td data-name="true">
    <img src="data:image/svg+xml;charset=UTF-8,%3Csvg%20width%3D%2232%22%20height%3D%2232%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%2032%2032%22%20preserveAspectRatio%3D%22none%22%3E%3Cdefs%3E%3Cstyle%20type%3D%22text%2Fcss%22%3E%23holder_18f57f0a593%20text%20%7B%20fill%3A%23FFFFFF%3Bfont-weight%3Abold%3Bfont-family%3AArial%2C%20Helvetica%2C%20Open%20Sans%2C%20sans-serif%2C%20monospace%3Bfont-size%3A10pt%20%7D%20%3C%2Fstyle%3E%3C%2Fdefs%3E%3Cg%20id%3D%22holder_18f57f0a593%22%3E%3Crect%20width%3D%2232%22%20height%3D%2232%22%20fill%3D%22%233a5a97%22%3E%3C%2Frect%3E%3Cg%3E%3Ctext%20x%3D%22-2.5390625%22%20y%3D%2228%22%3E32x32%3C%2Ftext%3E%3C%2Fg%3E%3C%2Fg%3E%3C%2Fsvg%3E"
         alt="Crypto Currency"
         class
    >
    <span>
      bitcoin
    </span>
    <span>
      BTC
    </span>
  </td>
  <td data-price="true">
    <span>
      $51,730.16
    </span>
  </td>
  <td data-changein1h="true">
    <span data-positive="false"
          data-negative="true"
    >
      -0.18%
    </span>
  </td>
  <td data-changein24h="true">
    <span data-positive="false"
          data-negative="true"
    >
      -0.13%
    </span>
  </td>
  <td data-changein1w="true">
    <span data-positive="true"
          data-negative="false"
    >
      9.5%
    </span>
  </td>
  <td data-marketcap="true">
    <span>
      $1.01T
    </span>
  </td>
  <td data-volume24h="true">
    <span>
      $23.43B
    </span>
  </td>
</tr>
`;

exports[`stories/components/CryptoCurrencyRow Ethereum smoke-test 1`] = `
<tr class="_container_y2li0_1">
  <td data-rank="true">
    <span>
      2
    </span>
  </td>
  <td data-name="true">
    <img src="data:image/svg+xml;charset=UTF-8,%3Csvg%20width%3D%2232%22%20height%3D%2232%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%2032%2032%22%20preserveAspectRatio%3D%22none%22%3E%3Cdefs%3E%3Cstyle%20type%3D%22text%2Fcss%22%3E%23holder_18f57f0a593%20text%20%7B%20fill%3A%23FFFFFF%3Bfont-weight%3Abold%3Bfont-family%3AArial%2C%20Helvetica%2C%20Open%20Sans%2C%20sans-serif%2C%20monospace%3Bfont-size%3A10pt%20%7D%20%3C%2Fstyle%3E%3C%2Fdefs%3E%3Cg%20id%3D%22holder_18f57f0a593%22%3E%3Crect%20width%3D%2232%22%20height%3D%2232%22%20fill%3D%22%233a5a97%22%3E%3C%2Frect%3E%3Cg%3E%3Ctext%20x%3D%22-2.5390625%22%20y%3D%2228%22%3E32x32%3C%2Ftext%3E%3C%2Fg%3E%3C%2Fg%3E%3C%2Fsvg%3E"
         alt="Crypto Currency"
         class
    >
    <span>
      ethereum
    </span>
    <span>
      ETH
    </span>
  </td>
  <td data-price="true">
    <span>
      $3,778.97
    </span>
  </td>
  <td data-changein1h="true">
    <span data-positive="false"
          data-negative="true"
    >
      -0.18%
    </span>
  </td>
  <td data-changein24h="true">
    <span data-positive="false"
          data-negative="true"
    >
      -0.13%
    </span>
  </td>
  <td data-changein1w="true">
    <span data-positive="true"
          data-negative="false"
    >
      9.5%
    </span>
  </td>
  <td data-marketcap="true">
    <span>
      $473.04M
    </span>
  </td>
  <td data-volume24h="true">
    <span>
      $36.24M
    </span>
  </td>
</tr>
`;
